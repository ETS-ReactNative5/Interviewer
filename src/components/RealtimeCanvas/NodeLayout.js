import React, { useRef, useEffect, useContext } from 'react';
import PropTypes from 'prop-types';
import LayoutContext from '../../contexts/LayoutContext';
import useDrag from './useDrag';
import LayoutNode from './LayoutNode';
// import getAbsoluteBoundingRect from '../../utils/getAbsoluteBoundingRect';

const LABEL = '0e75ec18-2cb1-4606-9f18-034d28b07c19';
const LAYOUT = 'd13ca72d-aefe-4f48-841d-09f020e0e988';

const NodeLayout = React.forwardRef(({
  // nodes,
  // allowPositioning,
  // highlightAttribute,
  // connectFrom,
  // allowSelect,
  onSelected,
  // layoutVariable,
  // width,
  // height,
}, oref) => {
  const ref = useRef();
  const layoutNodes = useRef([]);
  const { forceSimulation, viewport, nodes } = useContext(LayoutContext);
  const [getDelta, handleDragStart, handleDragMove, handleDragEnd] = useDrag();

  const handleClick = (e, index) => {
    onSelected(nodes[index]);
  };

  const timer = useRef();

  const update = useRef(() => {
    const {
      hasMoved,
      id,
      dy,
      dx,
    } = getDelta();

    if (id && hasMoved) {
      forceSimulation.updateNodeByDelta({ dy, dx }, id);
    }

    if (forceSimulation.simulation.current.nodes) {
      forceSimulation.simulation.current.nodes.forEach((position, index) => {
        const screenPosition = viewport.calculateViewportScreenCoords(
          viewport.calculateRelativeCoords(position),
        );

        layoutNodes.current[index].el.style.left = `${screenPosition.x}px`;
        layoutNodes.current[index].el.style.top = `${screenPosition.y}px`;
      });
    }

    // debugger;

    timer.current = requestAnimationFrame(() => update.current());
    // timer.current = setTimeout(() => update.current(), 100);
  });

  useEffect(() => {
    if (!ref.current) { return () => {}; }

    // debugger;

    const container = document.createElement('div');
    container.className = 'nodes-layout';
    container.setAttribute('id', 'nodes_layout');
    container.setAttribute('style', 'position: absolute; top: 0; left: 0; width: 100%; height: 100%; z-index: 200;');

    ref.current.appendChild(container);

    viewport.measureCanvas(container);

    return () => {
      debugger;
      ref.current.removeChild(container);
    };
  }, []);

  useEffect(() => {
    if (!ref.current) { return () => cancelAnimationFrame(timer.current); }
    // if (!ref.current) { return () => clearTimeout(timer.current); }

    // debugger;

    const container = document.getElementById('nodes_layout');

    layoutNodes.current = nodes.map((n) => {
      const a = n.attributes;

      return LayoutNode({
        color: 'color-neon-coral',
        label: a[LABEL],
      });
    });

    const els = layoutNodes.current.map(({ el }) => el);

    els.forEach((el, index) => {
      container.appendChild(el);

      el.addEventListener('mousedown', (e) => handleDragStart(e, index));
      el.addEventListener('click', (e) => handleClick(e, index));
    });

    window.addEventListener('mousemove', handleDragMove);
    window.addEventListener('mouseup', handleDragEnd);

    timer.current = requestAnimationFrame(() => update.current());
    // timer.current = setTimeout(() => update.current(), 100);

    return () => {
      // debugger;
      els.forEach((el) => container.removeChild(el));

      window.removeEventListener('mousemove', handleDragMove);
      window.removeEventListener('mouseup', handleDragEnd);

      cancelAnimationFrame(timer.current);
      // clearTimeout(timer.current);
    };
  }, [nodes]);

  return (
    <>
      <div className="node-layout" ref={ref} />
      <div style={{ position: 'absolute', top: 0, left: 0 }}>
        <button type="button" onClick={() => viewport.zoomViewport(1.5)}>in</button>
        <button type="button" onClick={() => viewport.zoomViewport(0.67)}>out</button>
        <button type="button" onClick={() => viewport.moveViewport(-0.1, 0)}>left</button>
        <button type="button" onClick={() => viewport.moveViewport(0, 0.1)}>down</button>
        <button type="button" onClick={() => viewport.moveViewport(0, -0.1)}>up</button>
        <button type="button" onClick={() => viewport.moveViewport(0.1, 0)}>right</button>
      </div>
      <div ref={oref} />
    </>
  );
});

NodeLayout.propTypes = {
  nodes: PropTypes.array.isRequired,
  onSelected: PropTypes.func.isRequired,
  allowPositioning: PropTypes.bool.isRequired,
  allowSelect: PropTypes.bool,
};

NodeLayout.defaultProps = {
  allowPositioning: true,
  allowSelect: true,
};

export { NodeLayout };

export default NodeLayout;
