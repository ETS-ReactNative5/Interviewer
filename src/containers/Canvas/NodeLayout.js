import React, { useState, useCallback, useRef } from 'react';
import { isNil } from 'lodash';
import { connect } from 'react-redux';
import { compose } from 'redux';
import { actionCreators as sessionsActions } from '../../ducks/modules/sessions';
import { entityPrimaryKeyProperty, entityAttributesProperty } from '../../ducks/modules/network';
// import { DropTarget } from '../../behaviours/DragAndDrop';
import NodeLayoutComponent from '../../components/Canvas/NodeLayout';
import useSize from '../../hooks/useSize';

const relativeCoords = (container, node) => ({
  x: (node.x - container.x) / container.width,
  y: (node.y - container.y) / container.height,
});

const accepts = ({ meta }) => meta.itemType === 'POSITIONED_NODE';

const NodeLayout = (props) => {
  const {
    onChange,
    createEdge,
    updateNode,
    toggleEdge,
    allowHighlighting,
    highlightAttribute,
    toggleHighlight,
    layoutVariable,
  } = props;

  const ref = useRef();
  const size = useSize(ref);

  const [connectFrom, setConnectFrom] = useState(null);
  const [rerenderCount, setRerenderCount] = useState(null);
  // Horrible hack for performance (only re-render nodes on drop, not on drag)
  const incrementRerenderCount = useCallback(() => setRerenderCount((c) => c + 1), []);
  const resetConnectFrom = useCallback(() => setConnectFrom(null), []);

  const onDrop = (item) => {
    const id = item.meta[entityPrimaryKeyProperty];
    const {
      width,
      height,
      x,
      y,
    } = size.current;

    updateNode(
      id,
      {},
      {
        [layoutVariable]: relativeCoords(
          {
            width, height, x, y,
          },
          item,
        ),
      },
    );

    onChange({
      type: 'node',
      id,
    });

    incrementRerenderCount();
  };

  const onDrag = (item) => {
    if (isNil(item.meta[entityAttributesProperty][layoutVariable])) { return; }
    const id = item.meta[entityPrimaryKeyProperty];
    const {
      width,
      height,
      x,
      y,
    } = size.current;
    updateNode(
      id,
      {},
      {
        [layoutVariable]: relativeCoords({
          width, height, x, y,
        }, item),
      },
    );

    onChange({
      type: 'node',
      id,
    });
  };

  const onDragEnd = () => {
    // make sure to also re-render nodes that were updated on drag end
    incrementRerenderCount();
  };

  const connectNode = useCallback((nodeId) => {
    // If edge creation is disabled, return
    if (!createEdge) { return; }

    // If the target and source node are the same, deselect
    if (connectFrom === nodeId) {
      resetConnectFrom();
      return;
    }

    // If there isn't a target node yet, set the selected node into the linking state
    if (!connectFrom) {
      setConnectFrom(nodeId);
      return;
    }

    // Either add or remove an edge
    if (connectFrom !== nodeId) {
      toggleEdge({
        from: connectFrom,
        to: nodeId,
        type: createEdge,
      });

      onChange({
        type: 'edge',
        from: connectFrom,
        to: nodeId,
      });
    }

    // Reset the node linking state
    resetConnectFrom();
  }, [createEdge, connectFrom, resetConnectFrom, setConnectFrom, toggleEdge]);

  const toggleHighlightAttribute = useCallback((node) => {
    if (!allowHighlighting) { return; }

    const newVal = !node[entityAttributesProperty][highlightAttribute];

    toggleHighlight(
      node[entityPrimaryKeyProperty],
      { [highlightAttribute]: newVal },
    );
  }, [allowHighlighting, highlightAttribute, toggleHighlight]);

  const onSelected = useCallback((node) => {
    console.log({ node, allowHighlighting });
    if (!allowHighlighting) {
      connectNode(node[entityPrimaryKeyProperty]);
    } else {
      toggleHighlightAttribute(node);
    }
    incrementRerenderCount();
  }, [allowHighlighting, incrementRerenderCount, connectNode, toggleHighlightAttribute]);

  return (
    <NodeLayoutComponent
      accepts={accepts}
      onDrag={onDrag}
      onDragEnd={onDragEnd}
      onDrop={onDrop}
      onSelected={onSelected}
      rerenderCount={rerenderCount}
      connectFrom={connectFrom}
      ref={ref}
      {...props}
    />
  );
};

const mapDispatchToProps = {
  toggleHighlight: sessionsActions.toggleNodeAttributes,
  toggleEdge: sessionsActions.toggleEdge,
  updateNode: sessionsActions.updateNode,
};

export default compose(
  connect(null, mapDispatchToProps),
)(NodeLayout);
