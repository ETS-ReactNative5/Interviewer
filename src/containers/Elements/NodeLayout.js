import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { bindActionCreators, compose } from 'redux';
import { connect } from 'react-redux';
import { isEqual, isMatch, filter, has, omit } from 'lodash';
import { createSelector } from 'reselect';
import LayoutNode from '../LayoutNode';
import { withBounds } from '../../behaviours';
import { DropZone } from '../../components/Elements';
import { makeNetworkNodesOfStageType } from '../../selectors/interface';
import { actionCreators as networkActions } from '../../ducks/modules/network';

const draggableType = 'POSITIONED_NODE';

const propsChangedExcludingNodes = (nextProps, props) =>
  !isEqual(omit(nextProps, 'nodes'), omit(props, 'nodes'));

const nodesLengthChanged = (nextProps, props) =>
  nextProps.nodes.length !== props.nodes.length;

class NodeLayout extends Component {
  constructor(props) {
    super(props);

    this.state = {
      connectFrom: null,
    };
  }

  shouldComponentUpdate(nextProps) {
    if (nodesLengthChanged(nextProps, this.props)) { return true; }
    if (propsChangedExcludingNodes(nextProps, this.props)) { return true; }

    return false;
  }

  onDropNode = () => {
    this.forceUpdate();
  }

  onSelectNode = (node) => {
    const { select } = this.props;

    if (!select && !select.action) { return; }

    switch (select.action) {
      case 'EDGE':
        this.connectNode(node.id); break;
      case 'ATTRIBUTES':
        this.toggleNodeAttributes(node.uid); break;
      default:
    }

    this.forceUpdate();
  }

  connectNode(nodeId) {
    const edgeType = this.props.edge.type;
    const connectFrom = this.state.connectFrom;

    if (!connectFrom) {
      this.setState({ connectFrom: nodeId });
      return;
    }

    if (connectFrom !== nodeId) {
      this.props.toggleEdge({
        from: connectFrom,
        to: nodeId,
        type: edgeType,
      });
    }

    this.setState({ connectFrom: null });
  }

  toggleNodeAttributes(nodeId) {
    this.props.toggleNodeAttributes({ uid: nodeId }, this.props.attributes);
  }

  isSelected(node) {
    const { select, canSelect } = this.props;
    if (!canSelect) { return false; }
    switch (select.action) {
      case 'EDGE':
        return (node.id === this.state.connectFrom);
      case 'ATTRIBUTES':
        return isMatch(node, this.props.attributes);
      default:
        return false;
    }
  }

  render() {
    const {
      layout,
      nodes,
      canPosition,
      canSelect,
    } = this.props;

    return (
      <DropZone droppableName="NODE_LAYOUT" acceptsDraggableType={draggableType}>
        <div className="node-layout">
          { nodes.map((node) => {
            if (!Object.prototype.hasOwnProperty.call(node, layout)) { return null; }

            return (
              <LayoutNode
                key={node.uid}
                layout={layout}
                node={node}
                draggableType={draggableType}
                onSelected={this.onSelectNode}
                onDropped={this.onDropNode}
                selected={this.isSelected(node)}
                canDrag={canPosition}
                canSelect={canSelect}
                areaWidth={this.props.width}
                areaHeight={this.props.height}
                animate={false}
              />
            );
          }) }
        </div>
      </DropZone>
    );
  }
}

NodeLayout.propTypes = {
  nodes: PropTypes.array,
  toggleEdge: PropTypes.func.isRequired,
  toggleNodeAttributes: PropTypes.func.isRequired,
  layout: PropTypes.string.isRequired,
  edge: PropTypes.object,
  select: PropTypes.object,
  attributes: PropTypes.object,
  canPosition: PropTypes.bool,
  canSelect: PropTypes.bool,
  width: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
};

NodeLayout.defaultProps = {
  nodes: [],
  attributes: {},
  edge: null,
  select: null,
  canPosition: false,
  canSelect: false,
};

const propLayout = (_, props) => props.prompt.sociogram.layout;

const makeGetPlacedNodes = () => {
  const networkNodesOfStageType = makeNetworkNodesOfStageType();

  return createSelector(
    [networkNodesOfStageType, propLayout],
    (nodes, layout) => filter(nodes, node => has(node, layout)),
  );
};

function makeMapStateToProps() {
  const getPlacedNodes = makeGetPlacedNodes();

  return function mapStateToProps(state, props) {
    const sociogram = props.prompt.sociogram;

    return {
      nodes: getPlacedNodes(state, props),
      layout: sociogram.layout,
      edge: sociogram.edge,
      select: sociogram.select,
      position: sociogram.position,
      attributes: sociogram.nodeAttributes,
      canPosition: !sociogram.select || !sociogram.select.action,
      canSelect: !!sociogram.select,
    };
  };
}

function mapDispatchToProps(dispatch) {
  return {
    toggleNodeAttributes: bindActionCreators(networkActions.toggleNodeAttributes, dispatch),
    toggleEdge: bindActionCreators(networkActions.toggleEdge, dispatch),
  };
}

export { NodeLayout as NodeLayoutPure };

export default compose(
  connect(makeMapStateToProps, mapDispatchToProps),
  withBounds,
)(NodeLayout);
