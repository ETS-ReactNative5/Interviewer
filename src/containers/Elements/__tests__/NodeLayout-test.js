/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { NodeLayoutPure } from '../NodeLayout';

const layout = 'foo';

const mockProps = {
  nodes: [
    { bar: 'buzz', uid: 123, [layout]: { x: 0, y: 0 } },
  ],
  updateNode: () => {},
  toggleEdge: () => {},
  toggleNodeAttributes: () => {},
  width: 123,
  height: 456,
  attributes: {},
  edge: null,
  select: null,
  canPosition: false,
  canSelect: false,
  layout,
};

describe('<NodeLayout />', () => {
  it('renders ok', () => {
    const component = shallow(<NodeLayoutPure {...mockProps} />);

    expect(component).toMatchSnapshot();
  });

  it('shouldComponentUpdate() when prop `nodes` changes length only', () => {
    const componentDidUpdate = jest.fn();
    NodeLayoutPure.prototype.componentDidUpdate = componentDidUpdate;

    const component = shallow(
      <NodeLayoutPure {...mockProps} />,
      { lifecycleExperimental: true },
    );

    component.setProps({
      ...mockProps,
      nodes: [{ bar: 'buzz', uid: 123 }],
    });

    component.setProps({
      ...mockProps,
      nodes: [{ bar: 'buzz', uid: 123 }, { bar: 'bing', uid: 456 }],
    });

    expect(componentDidUpdate.mock.calls.length).toEqual(1);
  });

  it('shouldComponentUpdate() when prop other than `nodes` changes', () => {
    const componentDidUpdate = jest.fn();
    NodeLayoutPure.prototype.componentDidUpdate = componentDidUpdate;

    const component = shallow(
      <NodeLayoutPure {...mockProps} />,
      { lifecycleExperimental: true },
    );

    component.setProps({
      ...mockProps,
      nodes: [{ bar: 'buzz', uid: 123 }],
    });

    component.setProps({
      ...mockProps,
      width: 789,
    });

    expect(componentDidUpdate.mock.calls.length).toEqual(1);
  });

  it('should forceUpdate() when onDropped callback is called', () => {
    const componentDidUpdate = jest.fn();
    NodeLayoutPure.prototype.componentDidUpdate = componentDidUpdate;

    const component = shallow(
      <NodeLayoutPure {...mockProps} />,
      { lifecycleExperimental: true },
    );

    component.find('Connect(LayoutNode)').at(0).prop('onDropped')();

    expect(componentDidUpdate.mock.calls.length).toEqual(1);
  });
});
