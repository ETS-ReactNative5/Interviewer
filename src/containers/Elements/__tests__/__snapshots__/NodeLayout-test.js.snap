// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NodeLayout /> renders ok 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Connect(Droppable)
    acceptsDraggableType="POSITIONED_NODE"
    droppableName="NODE_LAYOUT"
>
    <div
        className="node-layout"
    >
        <Connect(LayoutNode)
            animate={false}
            areaHeight={456}
            areaWidth={123}
            canDrag={false}
            canSelect={false}
            draggableType="POSITIONED_NODE"
            layout="foo"
            node={
                Object {
                    "bar": "buzz",
                    "foo": Object {
                      "x": 0,
                      "y": 0,
                    },
                    "uid": 123,
                  }
            }
            onDropped={[Function]}
            onSelected={[Function]}
            selected={false}
        />
    </div>
</Connect(Droppable)>,
  "nodes": Array [
    <Connect(Droppable)
      acceptsDraggableType="POSITIONED_NODE"
      droppableName="NODE_LAYOUT"
>
      <div
            className="node-layout"
      >
            <Connect(LayoutNode)
                  animate={false}
                  areaHeight={456}
                  areaWidth={123}
                  canDrag={false}
                  canSelect={false}
                  draggableType="POSITIONED_NODE"
                  layout="foo"
                  node={
                        Object {
                              "bar": "buzz",
                              "foo": Object {
                                "x": 0,
                                "y": 0,
                              },
                              "uid": 123,
                            }
                  }
                  onDropped={[Function]}
                  onSelected={[Function]}
                  selected={false}
            />
      </div>
</Connect(Droppable)>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <NodeLayout
        attributes={Object {}}
        canPosition={false}
        canSelect={false}
        edge={null}
        height={456}
        layout="foo"
        nodes={
                Array [
                        Object {
                          "bar": "buzz",
                          "foo": Object {
                            "x": 0,
                            "y": 0,
                          },
                          "uid": 123,
                        },
                      ]
        }
        select={null}
        toggleEdge={[Function]}
        toggleNodeAttributes={[Function]}
        updateNode={[Function]}
        width={123}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": NodeLayout {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "onDropNode": [Function],
        "onSelectNode": [Function],
        "props": Object {
          "attributes": Object {},
          "canPosition": false,
          "canSelect": false,
          "edge": null,
          "height": 456,
          "layout": "foo",
          "nodes": Array [
            Object {
              "bar": "buzz",
              "foo": Object {
                "x": 0,
                "y": 0,
              },
              "uid": 123,
            },
          ],
          "select": null,
          "toggleEdge": [Function],
          "toggleNodeAttributes": [Function],
          "updateNode": [Function],
          "width": 123,
        },
        "refs": Object {},
        "state": Object {
          "connectFrom": null,
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <Connect(Droppable)
          acceptsDraggableType="POSITIONED_NODE"
          droppableName="NODE_LAYOUT"
>
          <div
                    className="node-layout"
          >
                    <Connect(LayoutNode)
                              animate={false}
                              areaHeight={456}
                              areaWidth={123}
                              canDrag={false}
                              canSelect={false}
                              draggableType="POSITIONED_NODE"
                              layout="foo"
                              node={
                                        Object {
                                                  "bar": "buzz",
                                                  "foo": Object {
                                                    "x": 0,
                                                    "y": 0,
                                                  },
                                                  "uid": 123,
                                                }
                              }
                              onDropped={[Function]}
                              onSelected={[Function]}
                              selected={false}
                    />
          </div>
</Connect(Droppable)>,
        "_debugID": 2,
        "_renderedOutput": <Connect(Droppable)
          acceptsDraggableType="POSITIONED_NODE"
          droppableName="NODE_LAYOUT"
>
          <div
                    className="node-layout"
          >
                    <Connect(LayoutNode)
                              animate={false}
                              areaHeight={456}
                              areaWidth={123}
                              canDrag={false}
                              canSelect={false}
                              draggableType="POSITIONED_NODE"
                              layout="foo"
                              node={
                                        Object {
                                                  "bar": "buzz",
                                                  "foo": Object {
                                                    "x": 0,
                                                    "y": 0,
                                                  },
                                                  "uid": 123,
                                                }
                              }
                              onDropped={[Function]}
                              onSelected={[Function]}
                              selected={false}
                    />
          </div>
</Connect(Droppable)>,
      },
      "_renderedNodeType": 1,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <NodeLayout
    attributes={Object {}}
    canPosition={false}
    canSelect={false}
    edge={null}
    height={456}
    layout="foo"
    nodes={
        Array [
            Object {
              "bar": "buzz",
              "foo": Object {
                "x": 0,
                "y": 0,
              },
              "uid": 123,
            },
          ]
    }
    select={null}
    toggleEdge={[Function]}
    toggleNodeAttributes={[Function]}
    updateNode={[Function]}
    width={123}
/>,
}
`;
